src/runtime/tests/refiner-test.ts

```diff
@@ -17,338 +17,362 @@ import {Schema} from '../schema.js';
-        `);
-        ref = Refinement.fromAst(manifestAst[0].args[0].type.fields[0].type.refinement, typeData);
-        range = Range.fromExpression(ref.expression);
-        assert.isTrue(range.equals(new Range([Segment.closedClosed(10, 10)])));
-
-    }));
-    it('regression test for parse failure on operator ordering.', () => {
-      parse(`
+  it('Throws error when expression does not produce boolean result.', Flags.withFieldRefinementsAllowed(async () => {
+    const manifestAst = parse(`
+      particle Foo
+        input: reads Something {num: Number [ num + 5 ] }
+    `);
+    const typeData = {'num': 'Number'};
+    const ref = Refinement.fromAst(manifestAst[0].args[0].type.fields[0].type.refinement, typeData);
+    const data = {
+      num: 6,
+    };
+    assert.throws(() => {
+      ref.validateData(data);
+    }, `Refinement expression (num + 5) evaluated to a non-boolean type.`);
+  }));
+  it('Throws error when operators and operands are incompatible.', Flags.withFieldRefinementsAllowed(async () => {
+    const manifestAst = parse(`
+      particle Foo
+        input: reads Something {num: Number [ (num < 5) + 3 == 0 ] }
+    `);
+    const typeData = {'num': 'Number'};
+    assert.throws(() => {
+      Refinement.fromAst(manifestAst[0].args[0].type.fields[0].type.refinement, typeData);
+    }, `Refinement expression (num < 5) has type Boolean. Expected Number.`);
+  }));
+  it('Throws error when operators and operands are incompatible.', Flags.withFieldRefinementsAllowed(async () => {
+    const manifestAst = parse(`
+      particle Foo
+        input: reads Something {num: Number [ (num and 3) == 0 ] }
+    `);
+    const typeData = {'num': 'Number'};
+    assert.throws(() => {
+      Refinement.fromAst(manifestAst[0].args[0].type.fields[0].type.refinement, typeData);
+    }, `Refinement expression num has type Number. Expected Boolean.`);
+  }));
+  it('Throws error when operators and operands are incompatible.', Flags.withFieldRefinementsAllowed(async () => {
+    const manifestAst = parse(`
+      particle Foo
+        input: reads Something {name: Text [ name > 'Ragav' ] }
+    `);
+    const typeData = {'name': 'Text'};
+    assert.throws(() => {
+      Refinement.fromAst(manifestAst[0].args[0].type.fields[0].type.refinement, typeData);
+    }, `Refinement expression name has type Text. Expected Number.`);
+  }));
+  it('tests expression to range conversion.', Flags.withFieldRefinementsAllowed(async () => {
+    const manifestAst = parse(`
+      particle Foo
+        input: reads Something {num: Number [ ((num < 3) and (num > 0)) or (num == 5) ] }
+    `);
+    const typeData = {'num': 'Number'};
+    const ref = Refinement.fromAst(manifestAst[0].args[0].type.fields[0].type.refinement, typeData);
+    const range = Range.fromExpression(ref.expression);
+    assert.isTrue(range.equals(new Range([Segment.openOpen(0, 3), Segment.closedClosed(5, 5)])));
```

Comment:
Prefer a deep equality or other chai assertion over `isTrue`.

Reply:
hmm... giving this another look, feel free to ignore if this would be too complex -- it just dawned on me that this may get into the nuances our our type system.

Reply:
For the record, the .equals here is a Range specific function that is only superficially similar to chais' equality checking. Possibly due for a rename to avoid confusion.

Reply:
If it's performing an equality check between Range instances, then `.equals` is the right name. Some minor confusion over a test method is not really an issue.